// postgraphile/build/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const postgraphile_1 = require("./postgraphile");
exports.postgraphile = postgraphile_1.postgraphile;
exports.postgraphql = postgraphile_1.postgraphile;
exports.createPostGraphileSchema = postgraphile_1.createPostGraphileSchema;
exports.createPostGraphQLSchema = postgraphile_1.createPostGraphileSchema;
exports.watchPostGraphileSchema = postgraphile_1.watchPostGraphileSchema;
exports.watchPostGraphQLSchema = postgraphile_1.watchPostGraphileSchema;
exports.withPostGraphileContext = postgraphile_1.withPostGraphileContext;
exports.withPostGraphQLContext = postgraphile_1.withPostGraphileContext;
const pluginHook_1 = require("./postgraphile/pluginHook");
exports.makePluginHook = pluginHook_1.makePluginHook;
var postgraphile_core_1 = require("postgraphile-core");
exports.SchemaBuilder = postgraphile_core_1.SchemaBuilder;
exports.postGraphileBaseOverrides = postgraphile_core_1.postGraphileBaseOverrides;
exports.postGraphileClassicIdsOverrides = postgraphile_core_1.postGraphileClassicIdsOverrides;
exports.PostGraphileInflectionPlugin = postgraphile_core_1.PostGraphileInflectionPlugin;
exports.PostGraphileClassicIdsInflectionPlugin = postgraphile_core_1.PostGraphileClassicIdsInflectionPlugin;
exports.default = postgraphile_1.postgraphile;
